Blogs: Diffy integration, intro and detail

ygerasimov opened the pull request:
## The Issue

Draft of the announcement of Diffy integration. The blog post about the technical aspects of the integration.

I am still working on the illustration for the post.

tyler36 commented:
Thank you for the follow-up and creating the blog.

Is there a reason you didn't include the a link to video (https://www.loom.com/share/1944eb3462934279a175e65e16e715ed) you shared on Discord?

rfay commented:
Sorry I haven't reviewed yet, will get there!

There are a couple of complaints about minor linting issues in https://github.com/ddev/ddev.com/actions/runs/10164852537/job/28129244997?pr=232

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+  - Announcements
+---
+
+[Diffy](https://diffy.websote) is a visual regression testing tool that allows you to take screenshots of your website and compare them so you know exactly what pages got changed and where. We decided to build DDEV integration so it is possible to take screenshots from the local website, upload them to Diffy, and then compare them with screenshots from other environments.
```
```suggestion
[Diffy](https://diffy.website) is a visual regression testing tool that allows you to take screenshots of your website and compare them so you know exactly what pages got changed and where. We decided to build DDEV integration so it is possible to take screenshots from the local website, upload them to Diffy, and then compare them with screenshots from other environments.
```

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+Architecturally integration consists of:
+
+* Docker container with preinstalled node and Chrome 111
+* Nodejs app that runs Chrome
```
```suggestion
* Node.js app that runs Chrome
```

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+
+## App configuration
+App requires an API key from the service and also project ID. This is a pretty interesting setup that we borrowed from the Platformsh project.
+We keep the API key in the global configuration (i.e. `~/.ddev/global_config.yaml`) and project ID is in local configuration (your project’s config.yaml).
```
```suggestion
We keep the API key in the global configuration (i.e. `~/.ddev/global_config.yaml`) and project ID is in local configuration (your project’s `.ddev/config.yaml`).
```

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+App requires an API key from the service and also project ID. This is a pretty interesting setup that we borrowed from the Platformsh project.
+We keep the API key in the global configuration (i.e. `~/.ddev/global_config.yaml`) and project ID is in local configuration (your project’s config.yaml).
+
+First we check if variables are set and if they are not — we request the values (`read token`) and then save them with `ddev config global —web-environment-add` in case of global variable and `ddev config —web-environment-add` in case of project based variable.
```
```suggestion
First we check if variables are set and if they are not — we request the values (`read token`) and then save them with `ddev config global --web-environment-add` in case of global variable and `ddev config --web-environment-add` in case of project based variable.
```

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+
+Pay attention that ‘{{ }}’ are [go templates](https://ddev.readthedocs.io/en/stable/users/extend/additional-services/#template-action-replacements-advanced) and not bash code.
+But this is not the end — also, we need to copy the values of these variables into .env file for the app. So the app can access them.
+This is very non-trivial task that we got help from one of the maintainers Stas Zhuk (@stasadev). It has a lot of go template code, and we run it in [post_install_action](https://github.com/DiffyWebsite/ddev-diffy/blob/main/install.yaml#L55).
```
Better to use permalinks if you modify this file in the future.

```suggestion
This is very non-trivial task that we got help from one of the maintainers [Stas Zhuk](introducing-maintainer-stas.md). It has a lot of go template code, and we run it in [post_install_action](https://github.com/DiffyWebsite/ddev-diffy/blob/c6b1e2dedf6b1c5e186d357fd000ce1a568f70e5/install.yaml#L55).
```

stasadev replied:
It is better to use permalinks, if you change the add-on file in the future, the article links will remain in good condition.

```suggestion
This is very non-trivial task that we got help from one of the maintainers [Stas Zhuk](introducing-maintainer-stas.md). It has a lot of go template code, and we run it in [post_install_action](https://github.com/DiffyWebsite/ddev-diffy/blob/c6b1e2dedf6b1c5e186d357fd000ce1a568f70e5/install.yaml#L55).
```

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+## Docker container "diffy" user
+For security reasons it is important not to run all the processes in our container as root user. For that we create a user "diffy" and then run all the commands as this user. 
+
+This is accomplished by extending main container (see ["build" section](https://github.com/DiffyWebsite/ddev-diffy/blob/main/docker-compose.diffy.yaml#L7)) and here are [commands to create the user](https://github.com/DiffyWebsite/ddev-diffy/blob/main/diffy/Dockerfile).
```
```suggestion
This is accomplished by extending main container (see ["build" section](https://github.com/DiffyWebsite/ddev-diffy/blob/c6b1e2dedf6b1c5e186d357fd000ce1a568f70e5/docker-compose.diffy.yaml#L7-L13)) and here are [commands to create the user](https://github.com/DiffyWebsite/ddev-diffy/blob/main/diffy/Dockerfile).
```

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+title: "Diffy integration with DDEV"
+pubDate: 2024-08-03
+modifiedDate: 2024-08-03
+summary: My experience building an addon for DDEV
```
```suggestion
summary: My experience building an add-on for DDEV
```

stasadev reviewed (changes requested):
Nice!
I suggested little things to make it better.

ygerasimov commented:
@stasadev thank you very much for the review. All comments addressed.

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+
+Pay attention that ‘{{ }}’ are [go templates](https://ddev.readthedocs.io/en/stable/users/extend/additional-services/#template-action-replacements-advanced) and not bash code.
+But this is not the end — also, we need to copy the values of these variables into .env file for the app. So the app can access them.
+This is very non-trivial task that we got help from one of the maintainers [Stas Zhuk](introducing-maintainer-stas.md). It has a lot of go template code, and we run it in [post_install_action](https://github.com/DiffyWebsite/ddev-diffy/blob/c6b1e2dedf6b1c5e186d357fd000ce1a568f70e5/install.yaml#L55).
```
```suggestion
This is very non-trivial task that we got help from one of the maintainers [Stas Zhuk](introducing-maintainer-stas.md). It has a lot of go template code, and we run it in [post_install_actions](https://github.com/DiffyWebsite/ddev-diffy/blob/c6b1e2dedf6b1c5e186d357fd000ce1a568f70e5/install.yaml#L55).
```

stasadev commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+
+## Run npm install
+We want to run npm install inside of the container and not on the host. We accomplished it by adding `post-start` hook and specifying the service where to run the command.
+For that we created `config.diffy.yaml` and had instructions there. Pay attention that it is specified as `project_files` in install.yaml file.
```
```suggestion
For that we created `config.diffy.yaml` and had instructions there. Pay attention that it is specified as `project_files` in `install.yaml` file.
```

stasadev reviewed (approved):
Looks great to me, and two more small changes 🙂

rfay reviewed (changes requested):
I'm going to be a bit of a pain here. You've written a nice article about the internals of the add-on, but I don't think anybody will read that unless it's properly titled and has an introduction. Could it be called "Advanced Add-on: Internals of the ddev-diffy add-on" or something like that?

But I'd really like to have a blog before this that actually introduces the add-on, showing it in use, preferably with embedded video and narration.

Then this one could show the actual internals and emphasize the techniques you used.

I happen to note a couple of problems I see in the code; you're using `wget` and `unzip` expecting them to be available, and in many, many places they're not there by default. I'll take a look at the add-on and try it out. 

Let me know your reaction to this. I really do hate to be a pain, but your work is too valuable to be buried in internals-speak.

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+---
+title: "Diffy integration with DDEV"
```
```suggestion
title: "Diffy: Anatomy of an Advanced DDEV Add-on"
```

How about this title?

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,63 @@
+title: "Diffy integration with DDEV"
+pubDate: 2024-08-03
+modifiedDate: 2024-08-03
+summary: My experience building an add-on for DDEV
```
```suggestion
summary: My experience building an advanced add-on for DDEV
```

rfay reviewed (changes requested):
Some minor suggestions, including title. I didn't have adequate internet to actually try out the add-on yesterday, so will try to do that today.

ygerasimov commented:
@rfay I totally agree on the idea of having first blog post announcing the integration and demo how it works and then second one (this one) about internals. I just got too excited about technical implementation :). Let me write the proper announcement separately and we can publish it first. And then this one.

rfay commented:
It's fine either way. I do like the title I suggested, and of course need it to work when I test it, hoping for the workaround.

rfay commented on src/content/blog/ddev-diffy-introduction.md:
````diff
@@ -0,0 +1,66 @@
+To download the app we use `install.yaml` post_install_actions:
+```
+- rm -rf diffy-worker && mkdir diffy-worker
+- docker run -it --rm -v ./diffy-worker:/diffy-worker --user $DDEV_UID:$DDEV_GID ddev/ddev-utilities bash -c "cd /diffy-worker && wget -qO- https://github.com/DiffyWebsite/diffy-worker/archive/refs/heads/main.tar.gz | tar xz --strip-components=1"
````
Don't understand why you do this work instead of including this in the container. Is this intended to run on the host?

ygerasimov replied:
@rfay we do not want to include the code in the container as this code is used elsewhere as well. So it is better to keep it in a separate repository.

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,66 @@
+
+Watch [3 mins video](https://www.loom.com/share/1944eb3462934279a175e65e16e715ed) or read [documentation](https://docs.diffy.website/features/local-development/ddev-add-on).
+
+Architecturally integration consists of:
```
```suggestion
Architecturally, integration consists of:
```

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,66 @@
+  - TechNotes
+---
+
+[Diffy](https://diffy.website) is a visual regression testing tool that allows you to take screenshots of your website and compare them so you know exactly what pages got changed and where. We decided to build DDEV integration so it is possible to take screenshots from the local website, upload them to Diffy, and then compare them with screenshots from other environments.
```
```suggestion
[Diffy](https://diffy.website) is a visual regression testing tool that allows you to take screenshots of your website and compare them so you know exactly what pages got changed and where. We decided to build DDEV integration so it is possible to take screenshots from the local website, upload them to Diffy, and then compare them with screenshots from other environments.

The `ddev-diffy` add-on is at [github.com/DiffyWebsite/ddev-diffy](https://github.com/DiffyWebsite/ddev-diffy) and can be installed with the instructions there, including `ddev get DiffyWebsite/ddev-diffy`. It does require a Diffy API key to operate.
```

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,66 @@
+This is a particularly interesting approach as we use `ddev-utilities` to download the code and unpack it.
+
+## App configuration
+App requires an API key from the service and also project ID. This is a pretty interesting setup that we borrowed from the Platformsh project.
```
```suggestion
The app requires an API key and a project ID for Diffy. This interesting setup that was borrowed from the [ddev-platformsh](https://github.com/ddev/ddev-platformsh) project.
```

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,66 @@
+
+## App configuration
+App requires an API key from the service and also project ID. This is a pretty interesting setup that we borrowed from the Platformsh project.
+We keep the API key in the global configuration (i.e. `~/.ddev/global_config.yaml`) and project ID is in local configuration (your project’s `.ddev/config.yaml`).
```
As suggested in issue, you may want to just keep it in the diffy-worker .env if you only need it there.

ygerasimov replied:
Fixed. I removed the whole section of `App configuration`.

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,66 @@
+
+First we check if variables are set and if they are not — we request the values (`read token`) and then save them with `ddev config global --web-environment-add` in case of global variable and `ddev config --web-environment-add` in case of project based variable.
+
+Pay attention that ‘{{ }}’ are [go templates](https://ddev.readthedocs.io/en/stable/users/extend/additional-services/#template-action-replacements-advanced) and not bash code.
```
```suggestion
Note that ‘{{ }}’ are [go templates](https://ddev.readthedocs.io/en/stable/users/extend/additional-services/#template-action-replacements-advanced) and not bash code.
```

ygerasimov replied:
No longer relevant. I removed this section completely.

rfay commented on src/content/blog/ddev-diffy-introduction.md:
````diff
@@ -0,0 +1,66 @@
+```
+See [Makefile](https://github.com/DiffyWebsite/diffy-worker/blob/main/docker/Makefile) and [Readme](https://github.com/DiffyWebsite/diffy-worker) for more details.
+
+The biggest roadblock I had with the container is that default shell from ubuntu:22.04 is “dash”. So we had to replace it with bash to make node installed and used properly. Also pay attention that we install node with "n" so it is installed to `/usr/local/bin/node` so it is accessible for everyone.
````
Why would you use the quite non-standard `dash` in the first place? But if you wanted it, why didn't you `apt install dash` to get it? (I certainly think bash is a better choice)

ygerasimov replied:
I do not want to use `dash` but it is shipped with upstream container. We use bash everywhere.

rfay reviewed (changes requested):
This is looking great. Added some minor suggestions and random comments.

ygerasimov commented:
@rfay I have addressed your comments:
- moved existing post to "anatomy of advanced addon"
- removed the section about the configuration of the addon completely as now it is simple edits of `.env` file
- changed the copy of the introduction of Diffy post
- updated images

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: "Introduction of Diffy DDEV plugin"
```
```suggestion
title: "Introduction: The Diffy DDEV plugin"
```

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: "Introduction of Diffy DDEV plugin"
+pubDate: 2024-08-07
+modifiedDate: 2024-08-07
```
```suggestion
# modifiedDate: 2024-08-07
```

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: "Introduction of Diffy DDEV plugin"
+pubDate: 2024-08-07
```
Probably doesn't matter, but if you touch this you might as well update to the probable date.

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,30 @@
+featureImage:
+  src: /img/blog/2024/08/camera-and-laptop.jpeg
+  alt: DDEV and Diffy integration
+credit: Microsoft image creator
```
Doesn't matter, but people sometimes ask me what the prompt was.

rfay commented on src/content/blog/ddev-diffy-introduction.md:
```diff
@@ -0,0 +1,30 @@
+
+Once you have screenshots uploaded you can compare them to any other set of screenshots. For example with screenshots from production.
+
+Here is a [video walkthrough](https://www.loom.com/share/a3b750e32581458f9d2271969bba1bb8) for the comparing local environment with your production with the integration.
```
I would assume that you can do an embed, if you want to you can do that here.

rfay commented on src/content/blog/anatomy-advanced-ddev-addon.md:
```diff
@@ -0,0 +1,60 @@
+modifiedDate: 2024-08-07
+author: Yuri Gerasymov
+featureImage:
+  src: /img/blog/2024/08/dismantled-camera.jpeg
```
What a great picture!

rfay commented on src/content/blog/anatomy-advanced-ddev-addon.md:
```diff
@@ -0,0 +1,60 @@
+---
+title: "Diffy: Anatomy of an Advanced DDEV Add-on"
+pubDate: 2024-08-07
+summary: Lessons learned running DDEV on an inexpensive Chromebook.
```
I'm quite sure you wanted something else for the summary here?

rfay commented on src/content/blog/anatomy-advanced-ddev-addon.md:
```diff
@@ -0,0 +1,60 @@
+---
+title: "Diffy: Anatomy of an Advanced DDEV Add-on"
+pubDate: 2024-08-07
```
Doesn't matter, but update the date to what you think is the actual date.

rfay commented on src/content/blog/anatomy-advanced-ddev-addon.md:
```diff
@@ -0,0 +1,60 @@
+
+[Diffy](https://diffy.website) is a visual regression testing tool that allows you to take screenshots of your website and compare them so you know exactly what pages got changed and where. We decided to build DDEV integration so it is possible to take screenshots from the local website, upload them to Diffy, and then compare them with screenshots from other environments.
+
+The `ddev-diffy` add-on is at [github.com/DiffyWebsite/ddev-diffy](https://github.com/DiffyWebsite/ddev-diffy) and can be installed with the instructions there, including `ddev get DiffyWebsite/ddev-diffy`. It does require a Diffy API key to operate.
```
Consider linking to the other blog here.

rfay commented on src/content/blog/anatomy-advanced-ddev-addon.md:
```diff
@@ -0,0 +1,60 @@
+
+Let’s go through the way integration is built.
+
+## Download node app
```
People don't have to take any of these actions, true? Why is there a `docker run -it` here? 

I think you just do a `ddev get DiffyWebsite/diffy` right? If that's the whole thing, then say that.

rfay reviewed (changes requested):
Mighty close, thanks! The only main thing is the second advanced one should be clear that the install.yaml is doing the work, and explain what work it's doing.

ygerasimov commented:
@rfay I addressed the comments you provided.

Please edit the `pubDate` when you plan to publish these posts.

rfay reviewed (approved):
Let's go with it! Thanks!
